_posts/2020-06-01-spring-boot-hazelcast.md

```diff
@@ -0,0 +1,232 @@
+Before we go into the topic, how to use Hazelcast with Spring, let's have a look at the topologies of Hazelcast 
+cache cluster.
+
+There are two types of topologies
+ * topology with embedded cache,
+ * client-server topology.
+ 
+## Embedded Cache
+This type of topology means, that every instance of application has an integrated member. 
+
+![Embedded Cache Topology](/assets/img/posts/hazelcast/embedded-cache.png)
+
+In this case, the application and the cache data are running on one node. **When a new cache entry is written in
+the cache, `Hazelcast` takes care about the distribution to other members.** When data should be read from the cache, 
+then this data can be found on the same node, where the application is running.
+
+Let's have a look at how to build a cluster with an embedded `Hazelcast` cache topology and Spring application.
+`Hazelcast` supports many distributed data structures for caching. We will use a `Map` because it provides well known `get`
+and `put` operations. 
+
+First, we have to add the `Hazelcast` dependency. `Hazelcast` is just a Java library, so that can be done very easily.
+
+````groovy
+compile group: 'com.hazelcast', name: 'hazelcast', version: '4.0.1'
+````
+
+Now let's create a cache client for the application.
+
+```java
+@Component
+public class CacheClient {
+
+    public static final String CARS = "cars";
+    private final HazelcastInstance hazelcastInstance
+                            = Hazelcast.newHazelcastInstance();
+
+    public Car put(String number, Car car){
+        IMap<String, Car> map = hazelcastInstance.getMap(CARS);
+        return map.putIfAbsent(number, car);
+    }
+
+    public Car get(String key){
+        IMap<String, Car> map = hazelcastInstance.getMap(CARS);
+        return map.get(key);
+    }
+   
+   // other methods omitted
+
+}
+```
+That's it. Now the application has a distributed cache. The most important part of this code is the creation of
+a cluster member. It happens by calling the method `Hazelcast.newHazelcastInstance()`. The method `getMap()`
+creates a `Map` in the cache or returns an existing one. The only thing we have to do to set the name of the `Map`.
+When we want to scale our application, every new instance will create a new member and this member will join
+the cluster automatically.
+
+This approach has two advantages.
+ * It is very easy to set up the cluster,
+ * Data access is very fast.
+ 
+We don't need to set up a separate cache cluster. It means we can create a cluster very fast by adding a couple
+lines of code.
+
+If we want to read the data from the cluster, the data access is low-latency, because we don't need to send
+a request to the cache cluster over the net.
+
+But it brings drawbacks too. Imagine we have a system, that requires one hundred instances of our application.
+In this cluster topology, it means we would have one hundred of cluster members even though we don't need it.
```

Comment:
Can you add a sentence why we wouldn't need a cluster member for each application node? Is there another reason than memory consumption? And why would the synchronization be expensive? The expense would scale linearly with the number of nodes, wouldn't it?

Reply:
I explained it.